p2panda-net/src/sync/manager.rs

```diff
@@ -176,9 +185,29 @@ where
                     let peer = msg.peer;
                     let topic = msg.topic;
 
-                    // Keep track of all concrete topics we will be running sync sessions over.
-                    self.queue_sync_session(peer, topic).await;
-                    self.schedule_next_attempt().await?
+                    let sync_attempt = SyncAttempt::new(peer, topic);
+
+                    if let Err(err) = self.schedule_attempt(sync_attempt).await {
+                        // The attempt will fail if the sync queue is full, indicating that a high
+                        // volume of sync sessions are underway. In that case, we drop the attempt
+                        // completely. Another attempt will be scheduled when the next announcement of
+                        // this peer-topic combination is received from the network-wide gossip
+                        // overlay.
+                        error!("failed to schedule sync attempt: {}", err)
+                    }
+                }
+                _ = attempt_resync_interval.tick() => {
```

Comment:
I wonder if there's a better solution for this resync behaviour...

Right now we're using a simple `VecDeque`: every 5 seconds we pull an attempt from the queue and schedule it if the previous attempt was completed more than 60 seconds ago. If not, we push it to the back of the queue.

Is a 5 second interval frequent enough? What if we end up with 50 or 100 attempts in the queue? Do we need a solution that offers reattempt timing guarantees (e.g. reattempt will be made more than 60 seconds after the previous successful attempt, but not more than 70 seconds after)?

Reply:
The logic looks fine as-is to me right now.

"First ever sync attempts" should always have higher priority I believe and "> 1th attempts" lower priority, but I think this is currently given with the two separate "priority" channels (`inbox` vs. `resync_queue`).

It gets hard to reason about timings when there's a higher priority channel involved with unknown number of items, so I think it's fine to say we can't define the upper timing limit, only the minimum.